### PR TITLE
Migrate tracer_test to RSpec

### PR DESCRIPTION
**Pull request by marcotc**
This PR moves all test coverage provided by `test/tracer_test.rb` to either `spec/ddtrace/tracer_spec.rb` or `spec/ddtrace/span_spec.rb`.

Pretty much all tests were in a good state, and were either ported or already existed with the same semantics in `tracer_spec.rb`. One test (`test_trace_nil_resource)` fell under the Span responsibility and was migrated to `span_spec.rb`.

There are two Minitest files pending migration: `test/rate_by_service_sampler_test.rb`, a core test, and `test/contrib/rails/errors_test.rb`, a Rails test.

### CODE REVIEW

File: spec/ddtrace/tracer_spec.rb [hunk 1]
```diff
@@ -209,6 +278,11 @@
 
         it { expect(trace).to eq(result) }
 
+        it do
+          trace
+          expect(span.name).to eq(name)
+        end
```
Comment (ivoanjo) [inline spec/ddtrace/tracer_spec.rb:284]: I'd still love if we could reference helpers directly. I was searching through the source to find where a `let(:span)` was defined for this context, like we do for others, and had to attach `pry` to realize that this is `TracerHelpers#span`, not a `let`.

I know I may be somewhat annoying bring this up from time to time, but I really think we should avoid if possible these module inclusions that result in methods just showing up in random places. Especially in very complex contexts (such as this one) it makes the code really hard to follow without actually firing it up in a repl/debugger and seeing what's up.

File: spec/ddtrace/tracer_spec.rb [hunk 2]
```diff
@@ -253,12 +327,39 @@
       end
 
       context 'when nesting spans' do
-        it 'only the top most span has a runtime ID tag' do
+        it do
```
Comment (ivoanjo) [inline spec/ddtrace/tracer_spec.rb:330]: I've noticed that in this file we often use `it do` with quite complex assertions in the block. This is somewhat of an anti-pattern, because the automatic description that rspec generates is just nonsense:

````
$ bundle exec rspec spec/ddtrace/tracer_spec.rb:333
Datadog::Tracer
  #trace
    given a block
      when nesting spans
        is expected to eq "tag_2"
````

I suggest adding a description for the `it` to avoid this; for instance in this case we could use `it 'nests children spans created inside a parent span'` or something similar?

File: spec/ddtrace/tracer_spec.rb [hunk 3]
```diff
@@ -253,12 +327,39 @@
...
+          tracer.trace('parent', service: 'service-parent') do
+            tracer.trace('child1') { |s| s.set_tag('tag', 'tag_1') }
+            tracer.trace('child2', service: 'service-child2') { |s| s.set_tag('tag', 'tag_2') }
+          end
+
+          expect(spans).to have(3).items
+
+          child1, child2, parent = spans # Spans are sorted alphabetically by operation name
```
Comment (ivoanjo) [inline spec/ddtrace/tracer_spec.rb:338]: Minor: Have you considered encoding this note on the helper name? E.g. `spans_sorted_by_operation_name`?

**Review by ivoanjo — APPROVED** [review]
Left a few small notes, overall LGTM 👍

File: spec/ddtrace/tracer_spec.rb (hunk 1 above)
Comment (marcotc) [reply spec/ddtrace/tracer_spec.rb:284]: So `tracer`, `span`, and `spans` are the only "special" ones that I think exist in `ddtrace` that keep showing up all the time.

They are always available everywhere, and I think it makes sense for them to be always available by default, given their ubiquitous utility, and so far their abstraction has held up.

Also, if never invoked, they have no side-effect, so there's no ill in having them made available to all tests.

Do you think this is something that documentation would help with? Do you think that documenting this at the top of the `spec_helper.rb`, for example, would work, given it's more or less standard to assume that `spec_helper.rb` is always(™️) loaded for an RSpec test suite.

File: spec/ddtrace/tracer_spec.rb (hunk 2 above)
Comment (marcotc) [reply spec/ddtrace/tracer_spec.rb:330]: I don't recall deleting this comment, but yeah, I agree that making sure the RSpec output is clear and descriptive should always be the case. I'll address it.

File: spec/ddtrace/tracer_spec.rb (hunk 3 above)
Comment (marcotc) [reply spec/ddtrace/tracer_spec.rb:338]: `sorted_spans` maybe, given `spans` happens about 1 billion times in our test suite :)

File: spec/ddtrace/tracer_spec.rb (hunk 1 above)
Comment (ivoanjo) [reply spec/ddtrace/tracer_spec.rb]: > So tracer, span, and spans are the only "special" ones that I think exist in ddtrace that keep showing up all the time.

If it's just me, I can go away, but I do think it makes the codebase harder to comprehend from a newer joiner's point of view (mine, lol).

> Do you think this is something that documentation would help with? Do you think that documenting this at the top of the spec_helper.rb, for example, would work, given it's more or less standard to assume that spec_helper.rb is always(™️) loaded for an RSpec test suite.

Hmm I'm not sure having it in the `spec_helper.rb` would've helped. It did not occur to me at all to search there, so if that documentation existed already, I would've completely missed it.

My suggestion would be to make it a bit more explicit in every file that uses it, e.g. something like:

```ruby
RSpec.describe SomeTracerClass do
  let(:tracer) { TracerHelpers.tracer(self) }
  let(:span) { TracerHelpers.span(self) }
  let(:spans} { TracerHelpers.spans(self) }

   # ...
end
```

or less preferable but smaller using [the use symbols as metadata option](https://relishapp.com/rspec/rspec-core/docs/helper-methods/define-helper-methods-in-a-module#use-symbols-as-metadata):

```ruby
RSpec.describe SomeTracerClass, :'enable the #tracer, #span and #spans helpers'

end
```

e.g. the key is if something shows up somewhere, there should be a hint about it in the file source code.

**Comment by ivoanjo** [comment]
> Document all it dos in tracer_spec

❤️ ❤️ ❤️  Thanks! :)
